src/v/serde/serde.h

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "bytes/iobuf_parser.h"
+#include "reflection/type_traits.h"
+#include "serde/envelope_for_each_field.h"
+#include "serde/error.h"
+#include "serde/logger.h"
+#include "serde/type_str.h"
+#include "ssx/future-util.h"
+#include "utils/named_type.h"
+
+#include <iosfwd>
+#include <numeric>
+#include <string>
+#include <string_view>
+
+namespace serde {
+
+#define SERDE_TOKEN_PASTE(x, y) x##y
+#define SERDE_TOKEN_PASTE2(x, y) SERDE_TOKEN_PASTE(x, y)
+
+#define HAS_MEMBER_FN(name, ...)                                               \
+    template<typename T, typename = void>                                      \
+    struct SERDE_TOKEN_PASTE(help_has_, name)                                  \
+      : std::false_type {};                                                    \
+                                                                               \
+    template<typename T>                                                       \
+    struct SERDE_TOKEN_PASTE(                                                  \
+      help_has_,                                                               \
+      name)<T, std::void_t<decltype(std::declval<T>().name(__VA_ARGS__))>>     \
+      : std::true_type {};                                                     \
+                                                                               \
+    template<typename T>                                                       \
+    inline constexpr auto const SERDE_TOKEN_PASTE(                             \
+      has_, name) = SERDE_TOKEN_PASTE(help_has_, name)<T>::value;
+
+static auto iob = iobuf{};
+static auto iop = iobuf_parser{std::move(iob)};
+
+HAS_MEMBER_FN(serde_read, iop, version_t{0}, version_t{0}, size_t{0U})
```

Comment:
Hey ben could you clarify what you meant here? BTW, this code is completely rewritten without macros etc...in the new version so maybe a new review is the easiest thing to do.

Reply:
Just had a quick glance.  Essentially the pattern is:
```cpp
T t;
t.serde_read(args...);
```
I thought it might be useful to have a hook for:
```cpp
auto t{serde_construct<T>(args...);
```
Where `T` no longer requires a default constructor.

Reply:
Just to clarify: this would replace the (user implemented) `serde_read` member method, right?
So you would implement the global `serde_construct<T>` for your type (maybe in the serde namespace or it could work by ADL if you implement it in the namespace of the struct itself). But the write still happens through the member function serde_write()?

How about the asynchronous variants? I think the idea to eliminate the need for a default ctor is good. But I'm not sure on how to do it asynchronously. Maybe we need a `serde_construct_async<T>(args...) -> future<>` to replace `ss::do_with(T{}, [](T& t) { t.serde_read_async(args...).then(...) };`?

Is this a practical issue? Are there many types (that should go on the wire/disk) that don't have a default ctor and where it's hard/impossible to quickly write one?

Reply:
It might be possible to choose construction via constructor, or aggregate init as fallback 

We have clang-14 / full concepts now, if that simplifies the implementation.